unison-cli/src/Unison/Share/Sync.hs

```diff
@@ -326,112 +330,256 @@ dagbfs goal children =
+              -- nothing more for the dispatcher to do, when in fact a downloader thread just hasn't made it as
+              -- far as recording its own existence
+              recordWorking workerCount
+            forkDownloader (NESet.unsafeFromSet hashes1)
+            loop
+
+        -- Check to see if there are no hashes left to download, no outstanding workers, and no work in either queue
+        checkIfDoneMode :: STM (IO ())
+        checkIfDoneMode = do
+          workers <- readTVar workerCount
+          when (workers > 0) retry
+          isEmptyTQueue entitiesQueue >>= \case
+            False -> retry
+            True -> pure ()
+          isEmptyTQueue newTempEntitiesQueue >>= \case
+            False -> retry
+            True -> pure ()
+          pure (pure ())
+
+    -- Downloader thread: download entities, enqueue to `entitiesQueue`
+    downloader ::
+      TQueue (NESet Share.HashJWT, NEMap Hash32 (Share.Entity Text Hash32 Share.HashJWT)) ->
+      WorkerCount ->
+      NESet Share.HashJWT ->
+      IO ()
+    downloader entitiesQueue workerCount hashes = do
+      Share.DownloadEntitiesSuccess entities <-
+        httpDownloadEntities
+          httpClient
+          unisonShareUrl
+          Share.DownloadEntitiesRequest {repoName, hashes}
+      downloaded callbacks (NESet.size hashes)
+      atomically do
+        writeTQueue entitiesQueue (hashes, entities)
+        recordNotWorking workerCount
+
+    -- Inserter thread: dequeue from `entitiesQueue`, insert entities, enqueue to `newTempEntitiesQueue`
+    inserter ::
+      TQueue (NESet Share.HashJWT, NEMap Hash32 (Share.Entity Text Hash32 Share.HashJWT)) ->
+      TQueue (Set Share.HashJWT, Maybe (NESet Hash32)) ->
+      WorkerCount ->
+      IO Void
+    inserter entitiesQueue newTempEntitiesQueue workerCount =
+      connect \conn ->
+        forever do
+          (hashJwts, entities) <-
+            atomically do
+              entities <- readTQueue entitiesQueue
+              recordWorking workerCount
+              pure entities
+          newTempEntities0 <-
+            Sqlite.runTransaction conn do
+              NEMap.toList entities & foldMapM \(hash, entity) ->
+                upsertEntitySomewhere hash entity <&> \case
+                  Q.EntityInMainStorage -> Set.empty
+                  Q.EntityInTempStorage -> Set.singleton hash
+          atomically do
+            writeTQueue newTempEntitiesQueue (NESet.toSet hashJwts, NESet.nonEmptySet newTempEntities0)
+            recordNotWorking workerCount
+
+    -- Elaborator thread: dequeue from `newTempEntitiesQueue`, elaborate, "enqueue" to `hashesVar`
+    elaborator ::
+      TVar (Set Share.HashJWT) ->
+      TVar (Set Share.HashJWT) ->
+      TQueue (Set Share.HashJWT, Maybe (NESet Hash32)) ->
+      WorkerCount ->
+      IO Void
+    elaborator hashesVar uninsertedHashesVar newTempEntitiesQueue workerCount =
+      connect \conn ->
+        forever do
+            (join . atomically) do
```

Comment:
Is there a benefit to the `(join . atomically) do` pattern here over something like:

```
forever do
  mayNewTempEntities <- atomically $ do ...
  case mayNewTempEntities of
    Nothing -> ...
    Just _ -> ...
```

?

I find interleaving STM and IO makes it tougher to see what's being accomplished transactionally 😅

Reply:
Sure, we can switch it to not do the `join . atomically` thing. I think instead of a `Maybe (Set Hash)` (or w/e) I'll go with a one-off data type, so it's more clear what going down one branch versus another means.